contracts/0.4.24/Lido.sol

```diff
@@ -722,7 +590,171 @@ contract Lido is Versioned, StETHPermit, AragonApp {
+        if (_receiver == address(0)) revert("MINT_RECEIVER_ZERO_ADDRESS");
+        if (_amountOfShares == 0) revert("MINT_ZERO_AMOUNT_OF_SHARES");
+
+        _whenNotStakingPaused();
+
+        uint256 stethAmount = super.getPooledEthByShares(_amountOfShares);
+
+        uint256 newExternalBalance = EXTERNAL_BALANCE_POSITION.getStorageUint256().add(stethAmount);
+        uint256 maxExternalBalance = _getMaxExternalBalance();
+
+        require(newExternalBalance <= maxExternalBalance, "EXTERNAL_BALANCE_LIMIT_EXCEEDED");
+
+        EXTERNAL_BALANCE_POSITION.setStorageUint256(newExternalBalance);
+
+        mintShares(_receiver, _amountOfShares);
+
+        emit ExternalSharesMinted(_receiver, _amountOfShares, stethAmount);
+    }
+
+    /// @notice Burns external shares from a specified account
+    ///
+    /// @param _amountOfShares Amount of shares to burn
+    ///
+    /// @dev authentication goes through _isBurner() method
+    function burnExternalShares(uint256 _amountOfShares) external {
+        if (_amountOfShares == 0) revert("BURN_ZERO_AMOUNT_OF_SHARES");
+
+        _whenNotStakingPaused();
+
+        uint256 stethAmount = super.getPooledEthByShares(_amountOfShares);
+        uint256 extBalance = EXTERNAL_BALANCE_POSITION.getStorageUint256();
+
+        if (extBalance < stethAmount) revert("EXT_BALANCE_TOO_SMALL");
+
+        EXTERNAL_BALANCE_POSITION.setStorageUint256(extBalance - stethAmount);
+
+        burnShares(msg.sender, _amountOfShares);
+
+        emit ExternalSharesBurned(msg.sender, _amountOfShares, stethAmount);
+    }
+
+    /// @notice processes CL related state changes as a part of the report processing
+    /// @dev all data validation was done by Accounting and OracleReportSanityChecker
+    /// @param _reportTimestamp timestamp of the report
+    /// @param _preClValidators number of validators in the previous CL state (for event compatibility)
+    /// @param _reportClValidators number of validators in the current CL state
+    /// @param _reportClBalance total balance of the current CL state
+    /// @param _postExternalBalance total balance of the external balance
+    function processClStateUpdate(
+        uint256 _reportTimestamp,
+        uint256 _preClValidators,
+        uint256 _reportClValidators,
+        uint256 _reportClBalance,
+        uint256 _postExternalBalance
+    ) external {
+        _whenNotStopped();
+
+        _auth(getLidoLocator().accounting());
+
+        // Save the current CL balance and validators to
+        // calculate rewards on the next push
+        CL_VALIDATORS_POSITION.setStorageUint256(_reportClValidators);
+        CL_BALANCE_POSITION.setStorageUint256(_reportClBalance);
+        EXTERNAL_BALANCE_POSITION.setStorageUint256(_postExternalBalance);
+
+        emit CLValidatorsUpdated(_reportTimestamp, _preClValidators, _reportClValidators);
+        // cl and external balance change are logged in ETHDistributed event later
```

Comment:
not sure about external balance change though
this event already exists without this value